Doc/regions.md

```diff
@@ -0,0 +1,517 @@
+
+New objects are always created as a local object.
+If Veronapy features are not used, then all the objects in the system will be local objects and the runtime will behave as normal.
+
+### Regions
+
+Regions are used to pass mutable state between concurrent units of execution (behaviours).
+For usability, we need local references into the region, that are not part of the region itself.
+This is precisely what local objects are for as they can refer into the region.
+
+The core challenge is allowing local objects to refer into a region,
+while allowing the region to be sent to another thread.
+To do this, we need to track the number of references from local objects to a region, the "local reference count" (LRC).
+
+When we send a region to another behaviour we need to ensure that the LRC is zero.
+To make this more usable, if the LRC is not zero, we can scan the current frame to find all references to the region, and remove them.
+If there are still references, then we can scan all objects (from the existing doubly-linked list in the Python cycle detector) to find a reference,
+and report that as an error.
+
+When a local object is promoted to a region, we need to ensure that the LRC is correctly maintained.
+This involves adding new counts for reference to the newly added objects, and removing counts for edges that have become internal to the region.
+Consider the following example:
+
+```python
+x = {}
+#  x is a field of the frame object, which is local. 
+#  it points at a fresh dictionary object, which is also local.
+y = {}
+#  y is a field of the frame object, which is local. 
+#  it points at a fresh dictionary object, which is also local.
+x.f = y
+# x.f is a field of the dictionary object, which is local.
+# y is also local so no need to do anything in the write barrier.
+
+# Create a fresh region
+r = Region()
+# r is an region
+r.f = y
+# r.f is a field of the region object, which is in that region.
+# y was local, so we need to promote y to the region, r.
+# Before the assignment y has an RC of 2: One from the frame object, one from the object pointed to be x.
+# This adds two local RCs to the region.
+y = None
+# This reduces the localRC on r, by one. 
+x = None
+# This causes the old value of x to be collected
+# This in turn reduces the localRC on r by one as it clears the f field that points into the region.
+
+# We can now send the region r. It only has a single local RC r, and the operation will cause the 
+# local frame to clear that reference.
+send(r)
+# r is None at this point
+```
+
+> [Note that `send` is a placeholder operation that will be turned into proper concurrency support later.
+ It represents any operation that should prevent the region from being accessed by the current thread.]
+
+
+To re-establish the correct local reference count, we calculate the updates to LRC as follows.
+First we calculate the set of local objects, L, that are being made reachable from this region.
```

Comment:
Not important but I reflected earlier how this is very similar to Pony. 
I wonder if this scanning means that we can use Sebastian's distributed cycle detector scheme to manage cowns.